### PR TITLE
SSD1680 e-ink driver

### PR DESCRIPTION
**Pull request by adamkaliszan**
## Summary
E-Ink display SSD1680 support

## Impact
None

## Testing
Tested on EPS32 TTGO t5v2 with mono 2.9 e-ink display

### CODE REVIEW

File: boards/xtensa/esp32/ttgo_eink5_v2/compile_commands.json [hunk 1]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (Ouss4) [inline boards/xtensa/esp32/ttgo_eink5_v2/compile_commands.json:1]: Why do we need this file?

**Comment by Ouss4** [comment]
Can you please also rebase on top of the latest master, not sure why the tlsr8278adk80d/nsh is trying to enable custom board configs, the history of the file shows that the defconfig never had those... cc. @xiaoxiang781216

File: boards/xtensa/esp32/ttgo_eink5_v2/compile_commands.json (hunk 1 above)
Comment (adamkaliszan) [reply boards/xtensa/esp32/ttgo_eink5_v2/compile_commands.json:1]: No, I will remove it

File: boards/Kconfig [hunk 2]
```diff
@@ -1399,7 +1417,7 @@ config ARCH_BOARD_NUCLEO_L552ZE
 
 config ARCH_BOARD_QEMU_I486
 	bool "Qemu i486 Mode"
-	depends on ARCH_X86 || ARCH_I486
+	depends on ARCH_CHIP_QEMU_I486
```
Comment (xiaoxiang781216) [inline boards/Kconfig:1420]: The same change is merged https://github.com/apache/incubator-nuttx/pull/6374/files#diff-60cc096e3a9b22a769602cbbc3b0f5e7731e72db7b0338da04fcf665ed753b64R1402

**Comment by acassis** [comment]
> @acassis Thx for Your remarks. I was wondering if split it into two commits or no. The reason, why I decided to pull all together was testing possibility. I can split it into two commits. I have some questions
> 
>     1. Current implementation of driver was not tested on 3 color displays. Portrait mode with 3 colors should work. Landscape mode with 3 colors is not implemented (more complicated bit stripping). I decided to add #error macro in order to inform that some modes are not implemented.
>        What should I do with the code for 3 color displays. Remove it or leave in current state? Maybe I should remove options with 3 color displays in Kconfig file?
> 
>     2. Using E-Ink display has a little bit different philosophy in comparison to LCD or OLED. There is special command that redraw the screen. This operation is time consuming (about 1s) and energy consuming. Now display redraw content after changing the last row. I would like to add special IOCLT command or use the existing one, that as argument takes area. Area with width and height 0 determine, that we want to redraw screen. I didn't want to add driver that have an impact on using frame buffer. Is it acceptable in first version?
> 
>     3. Missing library for coping bits, like memcopy for coping the bytes. Now special functions are in driver file. Is it acceptable?
> 
>     4. There are 5 very similar functions ssd1680_snd_cmd_with_data0 .. ssd1680_snd_cmd_with_data4. I wanted to avoid using vargs macros. Is it acceptable or replace it by one function with variable number of arguments?

1. You can keep the entry in the Kconfig and just include ("not implemented yet").
2. Maybe you can support two options: an automatic redraw to use it similar to a LCD/OLED, then the last byte of framebuffer is written (last line/last column) is written it do the redraw and the IOCTL option to let user to redraw manually when some area of the display is written.
3. Yes, if you cannot use memcpy or similar, it is okay to have your own implementation.
4. I also thing it is better to have these 5 similar functions than using vargs, but there are some few drivers using vargs. If you decide to go this way, you are not the first one to follow this path.

File: boards/xtensa/esp32/common/Kconfig [hunk 3]
```diff
@@ -60,6 +60,34 @@ choice ESP32_SPIFLASH_FS
...
+config SSD1680_GPIO_PIN_CS
+	int "Pin that select the chip on SPI bus"
+	default "5" if ARCH_BOARD_TTGO_T5V2_ESP32
+
+config SSD1680_GPIO_PIN_DTA_CMD
+	int "Pin that switch between command and data on 4-wire SPI bus"
+	default "19" if ARCH_BOARD_TTGO_T5V2_ESP32
+
+config SSD1680_SPI_BUS
+	int "Spi bus no"
```
Comment (acassis) [inline boards/xtensa/esp32/common/Kconfig:85]:
```suggestion
	int "SPI Bus number"

**Comment by acassis** [comment]
@adamkaliszan seams there is an issue in the netnsh:
```
Configuration/Tool: ttgo_eink5_v2/netnsh
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Building NuttX...
src/esp32_ssd1680.c: In function 'board_lcd_initialize':
Error: src/esp32_ssd1680.c:137:9: error: implicit declaration of function 'esp32_spibus_initialize'; did you mean 'esp32_gpioirqinitialize'? [-Werror=implicit-function-declaration]
   spi = esp32_spibus_initialize(CONFIG_SSD1680_SPI_BUS);
         ^~~~~~~~~~~~~~~~~~~~~~~
         esp32_gpioirqinitialize
Error: src/esp32_ssd1680.c:137:7: error: assignment to 'struct spi_dev_s *' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
   spi = esp32_spibus_initialize(CONFIG_SSD1680_SPI_BUS);
       ^
cc1: all warnings being treated as errors
```

File: drivers/lcd/Kconfig [hunk 4]
```diff
@@ -323,38 +323,6 @@ config LCD_HILETGO
 		LCD_MAXCONTRAST should be 255, but any value >0 and <=255 will be accepted.
 		LCD_MAXPOWER should be 1:  0=off, 1=on
 
-config LCD_SSD1306_CUSTOM
```
Comment (xiaoxiang781216) [inline drivers/lcd/Kconfig:326]: why delete the custom section of ssd1306?

**Comment by xiaoxiang781216** [comment]
Please normalize your defconfig like this:
```
./tools/refresh.sh --silent ttgo_eink5_v2:fb
```
to fix the following ci fail:
```
HEAD detached at pull/6378/merge
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   boards/xtensa/esp32/ttgo_eink5_v2/configs/fb/defconfig
	modified:   boards/xtensa/esp32/ttgo_eink5_v2/configs/netnsh/defconfig
	modified:   boards/xtensa/esp32/ttgo_eink5_v2/configs/nsh/defconfig
```

File: boards/xtensa/esp32/common/Kconfig (hunk 3 above)
Comment (xiaoxiang781216) [reply boards/xtensa/esp32/common/Kconfig:85]: @adamkaliszan this comment isn't addressed.

**Comment by xiaoxiang781216** [comment]
@adamkaliszan CONFIG_LCD_SSD1680_2_9=y is redundant setting

**Comment by acassis** [comment]
@adamkaliszan your fb defconfig still incorrect, you need to Normalize it using the refresh.sh script, then you will see that this option was removed:
```
diff --git a/boards/xtensa/esp32/ttgo_eink5_v2/configs/fb/defconfig b/boards/xtensa/esp32/ttgo_eink5_v2/configs/fb/defconfig
index b4ef353f11..4eaa7f56d7 100644
--- a/boards/xtensa/esp32/ttgo_eink5_v2/configs/fb/defconfig
+++ b/boards/xtensa/esp32/ttgo_eink5_v2/configs/fb/defconfig
@@ -31,7 +31,6 @@ CONFIG_IDLETHREAD_STACKSIZE=3072
 CONFIG_INIT_ENTRYPOINT="nsh_main"
 CONFIG_INTELHEX_BINARY=y
 CONFIG_LCD_PORTRAIT=y
-CONFIG_LCD_SSD1680_2_9=y
 CONFIG_MM_REGIONS=3
 CONFIG_NSH_ARCHINIT=y
 CONFIG_NSH_BUILTIN_APPS=y
 ```

**Comment by adamkaliszan** [comment]
I will split this pull request int two parts: driver and BSP.

**Comment by adamkaliszan** [comment]
I would like to withdraw this pull request. I have added new pull request with driver only, without BSP

**Pull request closed without merging**